### DISCUSSION TITLE
middleware not working

### DISCUSSION THREAD

**Discussion by yun8boo** [vercel/next.js#38546] [Help]
Where should we put the middleware.ts when we put the pages directory under src?

version
- `next: 12.2.2`
- `typescript: 4.6.3`

directory
```
- .next
- node_modules
- src
  - pages
  - middleware.ts
```

ts.config
```ts.config
{
  "compilerOptions": {
    "target": "es5",
    "lib": ["dom", "dom.iterable", "esnext"],
    "allowJs": true,
    "skipLibCheck": true,
    "strict": true,
    "forceConsistentCasingInFileNames": true,
    "noEmit": true,
    "esModuleInterop": true,
    "module": "esnext",
    "moduleResolution": "node",
    "resolveJsonModule": true,
    "isolatedModules": true,
    "jsx": "preserve",
    "incremental": true,
    "baseUrl": "./",
    "paths": {
      "@/*": ["src/*"]
    }
  },
  "include": ["next-env.d.ts", "**/*.ts", "**/*.tsx"],
  "exclude": ["node_modules"]
}
```

**Comment by icyJoseph** [comment]
Hi,

It should be a sibling to `pages` so, `src/pages` and `src/middleware.ts` need to be at the same level.

What kind of issue are you seeing?

**Reply by yun8boo** [reply]
@icyJoseph 
It is placed at the same level as `src/pages`, but it is not working

middleware does not compile
![スクリーンショット 2022-07-13 9 26 20](https://user-images.githubusercontent.com/32152877/178623293-31cedff0-0b2c-45f6-9875-38fc1a5ce2d3.png)


`src/middleware.ts`
```ts
export function middleware(request: NextRequest) {
  console.log(request);
}
```

**Reply by yun8boo** [reply]
Adding `Page Extensions` to next.config.js stops middleware from working?

`next.config.js`
```ts
/** @type {import('next').NextConfig} */
const nextConfig = {
  reactStrictMode: true,
  swcMinify: true,
  pageExtensions: ['page.tsx', 'api.ts'],
};

module.exports = nextConfig;

```

**Reply by yun8boo** [reply]
I changed from `middleware.ts` to `middleware.api.ts` and it worked!

Thank you very much.

**Reply by icyJoseph** [reply]
Aha, this was a nice troubleshooting session. Kind of surprised that pageExtensions affects middleware.

**Reply by granmoe** [reply]
I had the same issue and agree that `pageExtensions` affecting middleware is surprising. Is there any drawback to making middleware.ts a "reserved" filename? Seems simpler and is similar to the conventions of the new layouts system with its many special filenames.

**Reply by gavmck** [reply]
@yun8boo Thank you, this fixed my issue too!

**Reply by araujobarret** [reply]
++ fixed my issue here, but I still consider it a bug from next-config

**Reply by DominicGBauer** [reply]
I made a PR to improve docs on this https://github.com/vercel/next.js/pull/44418

**Reply by christiankaindl** [reply]
> Adding Page Extensions to next.config.js stops middleware from working?

Thanks so much, was already loosing my mind ^^

**Reply by mitsue-eth** [reply]
Was losing my mind as well. Now all works. Thank you!

**Reply by alxdsz** [reply]
I have the middleware.ts placed correctly at the same level as pages but it seems to not work. The middleware is not triggered, I think it's not even taken into account on build. I'm going crazy with this, so will be more than thankful for any clues :)

**Reply by icyJoseph** [reply]
Btw, if you use `custom page extensions`, https://nextjs.org/docs/api-reference/next.config.js/custom-page-extensions, your `middlewate.ts` file has to be named, `middleware.page.ts`, or appropriately fit to your case.

**Reply by HelloCoovi** [reply]
> I have the middleware.ts placed correctly at the same level as pages but it seems to not work. The middleware is not triggered, I think it's not even taken into account on build. I'm going crazy with this, so will be more than thankful for any clues :)

This might be a little hard to understand, but I had the same problem with the `middleware.ts` file inside the src folder not running at all, but I solved it.
For reference, the version of Next.js I'm using is 13.3.0.

There are two ways to do this
Please choose and use one of the methods below.

1. Remove the `middleware.ts` file from the `src` folder and put it in the root directory.
2. remove the `pages` folder.

The pages folder used in the Next.js 12 version seems to be causing the conflict.
I suspect the problem is caused by mixing `app`, which is the new approach, with `pages`, which is the approach in version 12.

**Reply by fredriktre** [reply]
===========================
EDIT: 

I got it to work.

The name had to stay as middleware, and as I thought, it was the matcher.
The new matcher looks like this:
`export const config = {
  matcher: "/posts/admin/:path*",
};`

============
I have tried all of this, with no results...

I think I might just be stupid and the matcher is wrong, but idk.

Anyone who might have a clue?

The file is named middleware.ts
It's a sibling of pages.

(Btw, the code is kinda wonky as I have been trying to make it respond, but it should redirect either way, which it doesn't)
here is the code at least:
`import { NextResponse } from "next/server";
import type { NextRequest } from "next/server";
import { getIronSession } from "iron-session/edge";
import { sessionOptions } from "./lib/auth/session";

export async function middleware(req: NextRequest) {
  const res = NextResponse.next();
  const session = await getIronSession(req, res, sessionOptions);

  // do anything with session here:
  const { user } = session;

  console.log("from middleware", { user });
  
  return NextResponse.redirect(new URL('/', req.url));
//   return res;
};

export const config = {
  matcher: "/admin/",
};`

**Reply by BhathiyaPrasad** [reply]
**Yeah in my case i just put the middleware inside the src and it triggered thank you guys**

**Comment by pcibraro** [comment]
That's very unclear in the docs

Use the file middleware.ts (or .js) in the root of your project to define Middleware. For example, at the same level as pages or app, or inside src if applicable.

I would change that to make implicit that middleware should be sibling to pages or app. Also, that naming convention does not seem to work well when you want to have different middleware applied to different paths in your app.

**Comment by fjordskii** [comment]
> Btw, if you use `custom page extensions`, https://nextjs.org/docs/api-reference/next.config.js/custom-page-extensions, your `middlewate.ts` file has to be named, `middleware.page.ts`, or appropriately fit to your case.

true lifesaver🕺

**Comment by usama-tariq1** [comment]
Middleware is not working 

My structure

-app
 --page.js
 --layout.js
 --middleware.ts

I'm using the next 14.0.4 but the middleware is not working
this is my middleware

import { NextRequest, NextResponse } from "next/server";

export function middleware(req : NextRequest){
console.log('middleware hit')
return NextResponse.redirect("/")
}

**Reply by icyJoseph** [reply]
Shouldn't Middleware be at the same level as app? Move it up one level

**Reply by usama-tariq1** [reply]
Thanks @icyJoseph that work but how about if i want to have multiple middleware? perhaps different for other routes.

**Reply by icyJoseph** [reply]
Matchers and custom logic, because there can only be one Middleware.

**Reply by usama-tariq1** [reply]
Thanks @icyJoseph that helps. 👍

**Comment by Coderamrin** [comment]
I've implemented middleware it's working fine on dev enviornment also in local build
But it's not working on the live version which is on vercel
Strange!

**Reply by chev0004** [reply]
same, did you find solutions?

**Reply by Coderamrin** [reply]
yeah, I found a solution.
I had to update the NEXTAUTH_URL on vercel from localhost:3000 to the URL of the live site.

**Reply by chev0004** [reply]
thanks for the reply, I also found out last night that that was my problem too, my middleware was watching for localhost
